### PR TITLE
Add `insert_or_assign` for map

### LINKED ISSUE #317: [ENHANCEMENT]: Add `insert_or_assign` for `static_map`
### Is your feature request related to a problem? Please describe.

Depends on #314

Similar to `std::unordered_map::insert_or_assign`, we should add a new host function `static_map::insert_or_assign` that either inserts the kay-value pair or modifies the existing key with the given payload.

### Describe the solution you'd like

API reference: https://en.cppreference.com/w/cpp/container/unordered_map/insert_or_assign

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PointKernel**
Closes #317

This PR adds `static_map::insert_or_assign`.

It also removes the `[[nodiscard]]` qualifier for `retrieve_all` since they shouldn't be necessary.

### CODE REVIEW

**Comment by PointKernel** [comment]
@seunghwak Let me know if the current implementation can meet your needs.

File: include/cuco/detail/static_map/static_map_ref.inl [hunk 1]
```diff
@@ -230,6 +230,168 @@ class operator_impl<
...
+  __device__ constexpr bool attempt_insert_or_assign(value_type* slot,
+                                                     value_type const& value) noexcept
+  {
+    ref_type& ref_          = static_cast<ref_type&>(*this);
+    auto const expected_key = ref_.impl_.empty_slot_sentinel().first;
+
+    auto old_key      = ref_.impl_.compare_and_swap(&slot->first, expected_key, value.first);
+    auto* old_key_ptr = reinterpret_cast<key_type*>(&old_key);
+
+    // if key success or key was alreay present in the map
```
Comment (sleeepyjack) [inline include/cuco/detail/static_map/static_map_ref.inl]:
```suggestion
    // if key success or key was already present in the map
```

**Review by sleeepyjack — APPROVED** [review]
Looks good!

I have one question/concern: What happens if we concurrently issue a `insert_or_assign` and `erase` operation on the same key? Can the slot end up in an invalid state, e.g. `{erased_key_sentinel, updated_value}`?

**Comment by PointKernel** [comment]
> What happens if we concurrently issue a insert_or_assign and erase operation on the same key?

Is that even legal? :)

Need to check erase detail but IIRC it will reset payload value during erasure so the invalid state you mentioned above may never happen.

**Comment by sleeepyjack** [comment]
We could make it illegal to construct a device ref both with erase and insert_or_assign. That might be the safest solution. I can't even come up with a usecase for that combination.

**Comment by PointKernel** [comment]
> We could make it illegal to construct a device ref both with erase and insert_or_assign.

I don't worry much here but would be eager to learn the use case if a user reports a bug about this scenario.

**Pull request merged**